### PR TITLE
ARROW-12640: [C++] Fix errors from VS 2019 in cpp/src/parquet/types.h

### PR DESCRIPTION
**Pull request by ianmcook**
This resolves a problem where Visual Studio 2019 reported syntax errors in `cpp/src/parquet/types.h` after ARROW-11929

### CODE REVIEW

**Comment by ianmcook** [comment]
@github-actions crossbow submit test-build-vcpkg-win

File: cpp/src/parquet/types.h [hunk 1]
```diff
@@ -17,6 +17,16 @@
 
 #pragma once
 
+#ifdef _WIN32
+
+// parquet.thrift's OPTIONAL RepetitionType conflicts with a #define from
+// above, so we undefine it
+#ifdef OPTIONAL
+#undef OPTIONAL
+#endif
```
Comment (pitrou) [inline cpp/src/parquet/types.h:26]: Can you move this after the includes? Also, can you update the comment (there is no "define from above")?

Comment (ianmcook) [reply cpp/src/parquet/types.h]: Done in 1a77246

**Comment by ianmcook** [comment]
@github-actions crossbow submit test-build-vcpkg-win

**Review by pitrou — APPROVED** [review]
+1, thank you @ianmcook

**Comment by ianmcook** [comment]
Thank you @pitrou. I will merge after CI looks green

**Pull request closed without merging**